src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs

```diff
@@ -2727,15 +2733,410 @@ class Program
+                // (14,13): error CS8350: This combination of arguments to 'Program.F3(ref R)' is disallowed because it may expose variables referenced by parameter 'r3' outside of their declaration scope
+                //             F3(ref r1); // 1
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "F3(ref r1)").WithArguments("Program.F3(ref R)", "r3").WithLocation(14, 13),
+                // (14,20): error CS8168: Cannot return local 'r1' by reference because it is not a ref local
+                //             F3(ref r1); // 1
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "r1").WithArguments("r1").WithLocation(14, 20));
+        }
+
+        [Fact]
+        public void MethodArgumentsMustMatch_11()
+        {
+            var source =
+@"using System.Diagnostics.CodeAnalysis;
+ref struct R
+{
+    public R(ref int i) { }
+}
+class Program
+{
+    static void F1(bool b)
+    {
+        R x1 = default;
+        R y1 = default;
+        if (b)
+            F2(ref x1, ref y1); // 1
+        else
+            F2(ref y1, ref x1); // 2
+    }
+    static void F2(ref R x2, [UnscopedRef] ref R y2)
+    {
+    }
+}";
+            var comp = CreateCompilation(new[] { source, UnscopedRefAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (13,13): error CS8350: This combination of arguments to 'Program.F2(ref R, ref R)' is disallowed because it may expose variables referenced by parameter 'y2' outside of their declaration scope
+                //             F2(ref x1, ref y1); // 1
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "F2(ref x1, ref y1)").WithArguments("Program.F2(ref R, ref R)", "y2").WithLocation(13, 13),
+                // (13,28): error CS8168: Cannot return local 'y1' by reference because it is not a ref local
+                //             F2(ref x1, ref y1); // 1
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "y1").WithArguments("y1").WithLocation(13, 28),
+                // (15,13): error CS8350: This combination of arguments to 'Program.F2(ref R, ref R)' is disallowed because it may expose variables referenced by parameter 'y2' outside of their declaration scope
+                //             F2(ref y1, ref x1); // 2
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "F2(ref y1, ref x1)").WithArguments("Program.F2(ref R, ref R)", "y2").WithLocation(15, 13),
+                // (15,28): error CS8168: Cannot return local 'x1' by reference because it is not a ref local
+                //             F2(ref y1, ref x1); // 2
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "x1").WithArguments("x1").WithLocation(15, 28));
+        }
+
+        [Fact]
+        public void MethodArgumentsMustMatch_12()
+        {
+            var source =
+@"ref struct R
+{
+    public R(ref int i) { }
+}
+class Program
+{
+    static void F1(bool b)
+    {
+        R x1 = default;
+        int i = 42;
+        R y1 = new R(ref i);
```

Comment:
```suggestion
        scoped R y1 = new R(ref i);
```
Consider this to make it clearer what is happening.

Reply:
Changed this to a comment.